### ISSUE TITLE
[feature request] Support for --squash

### ISSUE THREAD

**Issue by Niek** [docker/buildx#727]
The `--squash` option is not yet implemented in buildx:
https://github.com/docker/buildx/blob/3dfbe2c18440f7c4a4f9643b093bd25b1ff65c37/commands/build.go#L85-L87

Are there any plans to add this soon?

**Comment by tonistiigi** [comment]
No plans atm. Squash was never taken out of experimental and is only supported in moby outside the builder component. You should squash layers with a multi-stage dockerfile(if you are sure you know what you are doing and need to squash at all).

**Comment by jfdenise** [comment]
I would vote for having squash in buildx, that is very convenient and reduces drastically the size of a an image.

**Comment by binoculars** [comment]
I'd also like to voice support for this issue. Even with multi-stage builds you still wind up with multiple layers that could be squashed into one. No need to download an untar a bunch of small files when one would do. Honestly, I'm hard pressed to think of an instance where a single layer and the end would not be preferred, assuming caching still works the same with squash.

**Comment by felipecrs** [comment]
If you have the ubuntu docker image pulled, and then you created a squashed image extended from ubuntu, when you try to pull it, it will pull it from scratch instead of reusing the existing Ubuntu image you have as a cached layer.

**Comment by tonistiigi** [comment]
> assuming caching still works the same with squash.

It does not work the same. Caching happens per layer, so you can't get any matches for squashed image.

Download would only be slower if you run something like `RUN rm ..` as part of your Dockerfile, again something that you should not do at all if you use multi-stage builds.

**Comment by binoculars** [comment]
Even with `mode=max`?

**Comment by jtougas** [comment]
I had been using `--squash` to reduce the size of the final image, and managed to get the equivalent reductions without it [thanks to this answer on SO](https://stackoverflow.com/questions/56117261/how-to-merge-dockers-layers-of-image-and-slim-down-the-image-file/56118557#56118557).

The idea is to add a final stage which copies in everything from the image:
```
FROM scratch as release
COPY --from=release-build / /
ENTRYPOINT ["/entrypoint.sh"]
```

**Comment by felipecrs** [comment]
Hm... interesting, but I think this will probably wipe out all environment variables and labels as well, right?

**Comment by binoculars** [comment]
> Hm... interesting, but I think this will probably wipe out all environment variables and labels as well, right?

Right. Can we just have squash for those people that really want to use it? I just want one layer at the end. For cases where you have one or two binaries, maybe some `ENV`s and bunch of `LABEL`s. I'd rather not have to do yet another `FROM` scratch and `COPY --from=previous / /`  It is faster because you're not waiting for multiple tiny layers to extract. It should be able to cache all of the intermediate layers, with `--cache-to mode=max`. Yes the final image is not cacheable, but if we're going from scratch anyway does it matter?

**Comment by tonistiigi** [comment]
> Even with mode=max?

`--squash` (the one still in dockerd) does not cache anything. If you use multi-stage builds then `mode=max` works but `mode=max` only works with a completely separate target where now all your layers (including the ones where you ran `rm ..` if you use the full `/` squash) are now pushed. For the regular inline cache that is based on the image itself(and does not make your build slower), you can't get any cache because there are no layers in the image.

**Comment by jsternberg** [comment]
I'm closing this because I don't think this is something we're ever going to work on and there are alternatives in a more buildkit native way to do this. If this is something people feel really strongly about, I can reopen this and convert it to a discussion or we can start a discussion to talk about what this workflow enables and how to make that workflow easy with buildkit.

**Comment by felipecrs** [comment]
Here's one idea, btw:

- https://github.com/moby/buildkit/issues/6062
